### ISSUE TITLE
percolate error

### ISSUE THREAD

**Issue by troonmel** [hklarner/pyboolnet#100]
Hi,
percolate() function in prime_implicants.py sometimes reach max_iterations before fully percolating even when using the default max_interations setting(which is more than the total number of nodes).
This leads to an error when the constants are removed.
pystablemotifs([https://github.com/jcrozum/pystablemotifs](https://github.com/jcrozum/pystablemotifs)) relies on this function, so it'll be very helpful if you take a look at this.

Below is a simple example. The error occurs most of the time.
```
from pyboolnet.external.bnet2primes import bnet_text2primes
from pyboolnet.prime_implicants import percolate

primes = bnet_text2primes("""A, A
B, A | B
C, B | C
D, C | D
E, D | E
F, E | F
G, F | G""")

percolate(primes, add_constants={'A':1}, remove_constants = True, copy=True)
```

I think this happens when there are many nodes with self-regulation. The node with self edge is added back into the searchspace of successors even after that node is fixed to a constant value. This sometimes prevents all nodes being visited within the finite number of iteration.

I tried adding a line that prevents the node that was fixed to a constant to be added back to the searchspace. I think this fixes the issue.

prime_implicants.py line 649 ~ line 657
```
    while successors and iterations < max_iterations:
        name = successors.pop()
        current_successors = set(find_successors(primes=primes, sources=[name]))
        update_primes(primes=primes, name=name, constants=constants)

        if is_constant(primes=primes, name=name):
            iterations += 1
            current_successors.discard(name) # new line that I added
            successors.update(current_successors)
            constants[name] = 1 if primes[name][1] else 0
```

Can you please take a look at this?

Thanks,

Kyu Hyong

**Comment by hklarner** [comment]
Hey Kyu,

I think you're right, thanks for reporting and suggesting a solution. Here is the PR, take a look::
- https://github.com/hklarner/pyboolnet/pull/101

**Comment by hklarner** [comment]
The fix is available in the new version 3.0.10:
- https://github.com/hklarner/pyboolnet/discussions/90#discussioncomment-3618925
